server/cronworkflow/cron_workflow_server.go

```diff
@@ -46,21 +70,74 @@ func (c *cronWorkflowServiceServer) GetCronWorkflow(ctx context.Context, req *cr
 	if req.GetOptions != nil {
 		options = *req.GetOptions
 	}
-	return auth.GetWfClient(ctx).ArgoprojV1alpha1().CronWorkflows(req.Namespace).Get(req.Name, options)
+	return c.getCronWorkflow(ctx, req.Namespace, req.Name, options)
 }
 
 func (c *cronWorkflowServiceServer) UpdateCronWorkflow(ctx context.Context, req *cronworkflowpkg.UpdateCronWorkflowRequest) (*v1alpha1.CronWorkflow, error) {
-	cronWf, err := auth.GetWfClient(ctx).ArgoprojV1alpha1().CronWorkflows(req.Namespace).Update(req.CronWorkflow)
+	_, err := c.getCronWorkflow(ctx, req.Namespace, req.Name, metav1.GetOptions{})
 	if err != nil {
 		return nil, err
 	}
-	return cronWf, nil
+	return auth.GetWfClient(ctx).ArgoprojV1alpha1().CronWorkflows(req.Namespace).Update(req.CronWorkflow)
 }
 
 func (c *cronWorkflowServiceServer) DeleteCronWorkflow(ctx context.Context, req *cronworkflowpkg.DeleteCronWorkflowRequest) (*cronworkflowpkg.CronWorkflowDeletedResponse, error) {
-	err := auth.GetWfClient(ctx).ArgoprojV1alpha1().CronWorkflows(req.Namespace).Delete(req.Name, req.DeleteOptions)
+	_, err := c.getCronWorkflow(ctx, req.Namespace, req.Name, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+	err = auth.GetWfClient(ctx).ArgoprojV1alpha1().CronWorkflows(req.Namespace).Delete(req.Name, req.DeleteOptions)
 	if err != nil {
 		return nil, err
 	}
 	return &cronworkflowpkg.CronWorkflowDeletedResponse{}, nil
 }
+
+func (c *cronWorkflowServiceServer) withInstanceID(opt metav1.ListOptions) metav1.ListOptions {
+	if c.mode == KubeClientMode {
```

Comment:
should also work for kube client - should not need the guard condition

Reply:
This is needed. 

The expected behaviors for `workflow list` on `GRPC server` and `kube-client` are different.

- On `GRPC server`, we expect it to list the workflows either no `instanceID` (instanceID == "") or `instanceID` is the same (intanceID != "").

- When it's `kube-client`, we want `argo list` to list all the workflows without any instanceID related `labelSelector`, without this we can not achieve that.

Reply:
I'm not sure about this. As much as possible, both clients should work in the same manner. 

This would mean that the GRPC server should be the same as the kube-client. 

Can you merge the labels together? 

I think you're doing great work btw.

Reply:
Thanks.

`argo server client` works the same as `GRPC server`, but I don't think `kube-client` does.

A simple question, what is expected behavior for `kube-client` if there are multiple workflow controllers with different `instanceid` installed in one cluster? As a `kube-client`, you can not be the same as any of the argo servers, instead of being limited to a subset of the objects (`intanceid` lableselector), by default you should have the ability to see whatever you want to see.

Reply:
No, I don't think we want that. The kube-client is legacy code really, so we should not make extra effort to maintain any esoteric features. If you need to make a choice here, can you do whatever is simpler. I can even accept breaking changes to the kube-client.

Reply:
If this is the case, let's pick up the legacy client code, if later on we decide not to support it any more, we can easily dump it.